README.md
[![License](https://img.shields.io/badge/License-GPL_v2-blue.svg?style=flat-square)](LICENSE)
[![made-with-javascript](https://img.shields.io/badge/Made%20with-JavaScript-1f425f.svg)](https://en.wikipedia.org/wiki/JavaScript)

# POMjs

**POMjs** is a random password generator in HTML and pure Javascript. It can be customized by modifying the sources. There's really nothing magic going on here, but it works and is somewhat responsive. The goal was to make something small, useful, and reasonably free from dependencies.

## Parts

**POMjs** uses a "minified" version of [normalize.css](https://necolas.github.io/normalize.css) to reset the DOM to reasonable settings :blush:

**POMjs** includes no externally loaded resources and contains no tracking code or cookie management.

## Installation

Download the distribution or clone the repo. Files and folders should be placed in your web root, or another folder accessible to your web server. All references to `CSS` and `Javascript` use relative paths.

There may be references to "Öppet Moln" ("Open Cloud"), which is a Swedish site run by the author of **POMjs**, [oppetmoln.se](https://oppetmoln.se), to promote open source solutions for general use. You can, of course, delete such references.

I'd appreciate a mention on whatever website you use this one, and a link to the original repo, but it's not required.

## Usage

* From a visitor perspective, simply enable the desired string sequences, and possibly adjust the password length. Click on the <kbd>Enter</kbd> button or use the <kbd>Enter</kbd> key when the button has focus.
* The generated password is displayed at the bottom of the box, along with a strength indicator.
* The slider can also be used to change the number of characters in the generated password, as can the input field to the right of the slider.
* The HTML should have proper `tabindex` attributes for easier keyboard navigation.
* To copy the generated password to the clipboard, click on the password.
* "Dark mode" (browser setting) should be honored.

## Changes

Please see [CHANGELOG.md](CHANGELOG.md) for a list of changes.

## Customization

The original Javascript source is in `password-om.js`. The `index.html` file includes `password-om.min.js`. So I suggest you make your customizations in the "unminified" .js file and then create a new `password-om.min.js` file. If you do not have access to a "minification tool", you can simply copy an "unminified" version of `password-om.js` to `password-om.min.js`, it's really not that big of a difference.

Most parameters/settings can be changed in the `.js` file.

### Variables (in .js file)

| Name                      | Default              | Comment                                    |
| ------------------------- | -------------------- | ------------------------------------------ |
| POM_strUppercase          | Auto-generated A-Z   | Change if needed, used by default          |
| POM_strLowercase          | Auto-generated a-z   | Change if needed, used by default          |
| POM_strDigits             | Auto-generated 0-9   | Change if needed, used by default          |
| POM_strSpecialOne         | -.\_#$@%!            | Change if needed, used by default          |
| POM_strSpecialTwo         | "+(){}[]?&,\*<>\|:;^ | Change if needed, **not** used by default  |
| POM_strUppercase_Default  | true                 | Enables POM_strUppercase by default        |
| POM_strLowercase_Default  | true                 | Enables POM_strLowercase by default        |
| POM_strDigits_Default     | true                 | Enables POM_strDigits by default           |
| POM_strSpecialOne_Default | true                 | Enables POM_strSpecialOne by default       |
| POM_strSpecialTwo_Default | **false**            | Enables POM_strSpecialTwo by default       |
| POM_mkPasswordOnLoad      | true                 | Generate password on first page load       |
| POM_changeIsClick         | true                 | Generate password when parameters are changed |
| POM_minLength             | 16                   | Minimum (and default) password length      |
| POM_maxLength             | 256                  | Maxmium password length                    |
| POM_sliderStep            | 8                    | 1 - ..                                     |
| POM_debug                 | false                | Emit "debug" output with console.log()     |

### CSS

Most of the styling is done via CSS. I have tried to make it reasonably responsive, and I have tested with it with various browsers.

Please note that `index.html` makes use of `password-om.min.css` by default. 

### Translation

The (very) few strings can be translated. The distribution contains strings in English (default), Swedish, German, Dutch, Finnish, French, and Luxembourgish. The code attempts to guess the language from the **navigator.language** properties. If they do not exist, or cannot be interpreted, the fallback language used is English.

The "string setup" is done in the `POM_initialSetup()` function.

If you would like to contribute to the translation, please open an issue and post the language and the translated strings, please include how/if you would like to be mentioned in the credits.

## Credits

Thanks to **Peter Hampf** (hello `O/T-Track` :metal:) for the German translation :blush:
Thanks to **Jeroen van de Leur** for the Dutch translation :blush:
Thanks to **Alain Fontaine** for the French and Luxembourgish translations :blush:
Thanks to **Thomas Raehalme** for the Finnish translations :blush:
Thanks to **Mihaly Balassy** for the Hungarian translations :blush:
Thanks to **Gregor Godler** for the Slovenian translations :blush:
Thanks for **Sarah Bailey** for assistance in "debugging" use on iOS/macOS :blush:

The "cloud icon" comes from [Streamline](https://streamlinehq.com)

**POMjs** was written by Joaquim Homrighausen while converting caffeine into code :coffee:

## License

**POMjs** is licensed under the GPLv2 license. See the [`LICENSE`](LICENSE) file for more details.

Copyright 2022, 2023 Joaquim Homrighausen.

### Demo

You can see POMjs in action, and more or less unmodified, on the following links:

* Öppet Moln ("Open Cloud"); [password.oppetmoln.se](https://password.oppetmoln.se)
* WebbPlatsen i Sverige AB; [password.webbplatsen.se](https://password.webbplatsen.se)

### External references

These links are not here for any sort of endorsement or marketing, they're purely for informational purposes :sunglasses:

* WebbPlatsen; [webbplatsen.se](https://webbplatsen.se)
* Öppet Moln ("Open Cloud"); [oppetmoln.se](https://oppetmoln.se)
* me; [joho.se](https://joho.se) and [github.com/joho1968](https://github.com/joho1968)
* POMjs repo; [github.com/joho1968/POMjs](https://github.com/joho1968/POMjs)
* normalize.css repo; [github.com/necolas/normalize.css](https://github.com/necolas/normalize.css)
* Streamline, Icons and Illustrations; [streamlinehq.com](https://streamlinehq.com)


CHANGELOG.md
# Changelog for POMjs
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.4] - 2023-07-25

### Added
- Added Slovenian (sl) translation (#6); thank you Gregor Godler
- Added missing Hungarian (hu) translation (#5) in meta description

### Changed
- Some minor typeface (font) changes to play better with iOS/macOS browsers
- Some minor UI changes to make password strength indicator more clear

## [1.1.3] - 2023-06-05

### Added
- Added Hungarian (hu) translation (#5); thank you Mihaly Balassy

## [1.1.2] - 2023-02-21

### Added
- Added Luxembourgish (lb) translation (#2); thank you Alain Fontaine
- Added French (fr) translation (#3); thank you Alain Fontaine
- Added Finnish (fi) translation (PR#3); thank you Thomas Raehalme

## [1.1.1] - 2023-01-23

### Added
- Added Dutch translation (#1); thank you Jeroen van de Leur

## [1.1.0] - 2022-10-24

### Changed
- Moved loading of the javascript file to `<head>` with `defer` attribute to be more adherent to standards.

### Added
- Added `CHANGELOG.md` :blush:
- Added links to "demo" sites using POMjs
- Added `autofocus` attribute to password generation `button`, this allows <kbd>Space></kbd> and <kbd>Enter</kbd> keys to be used to immediately generate passwords.
- Added HTML "outline" to the password generation button when it has focus to make it more obvious that is has focus
- Added automatic password generation (configurable) when parameters are changed, default is enabled.

### Removed
- Removed incorrect debug output (it would be logged even if debugging was disabled)

## [1.0.0] - 2022-10-21
- Initial release


js/password-om.js
/**
 * POMjs 1.1.4
 *
 * password-om.js
 *
 * Plain vanilla javascript for the "Öppet Moln" ("Open Cloud") random password
 * generator site (password.oppetmoln.se). For the sake of having a name, it
 * shall be POMjs. There's really nothing fancy going on here :-)
 *
 * Copyright 2022, 2023 Joaquim Homrighausen; All rights reserved.
 * Development sponsored by WebbPlatsen i Sverige AB
 * https://www.webbplatsen.se
 *
 * This file is part of POMjs. POMjs is free software.
 *
 * You may redistribute it and/or modify it under the terms of the
 * GNU General Public License version 2, as published by the Free Software
 * Foundation.
 *
 * POMjs is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License
 * along with the POMjs package. If not, write to:
 *  The Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor
 *  Boston, MA  02110-1301, USA.
 */

/* Some globals, to be filled by script */

var POM_cfgLanguage = 'noLang?';
var POM_cfgTitle = 'noTitle?';
var POM_cfgSlogan = 'noSlogan?';
var POM_cfgGenPasswordHint = 'noHint?';
const POM_strUppercase = POM_genAlphabet(true);
const POM_strLowercase = POM_genAlphabet(false);
const POM_strDigits = POM_genDigits();
const POM_strSpecialOne = '-._#$@%!';
const POM_strSpecialTwo = '"+(){}[]?&,*<>|:;^';

/* Easy customization of defaults, etc */
const POM_strUppercase_Default = true;
const POM_strLowercase_Default = true;
const POM_strDigits_Default = true;
const POM_strSpecialOne_Default = true;
const POM_strSpecialTwo_Default = false;
const POM_mkPasswordOnLoad = true;
const POM_changeIsClick = true;
const POM_minLength = 16;
const POM_maxLength = 256;
const POM_sliderStep = 8;
const POM_debug = false;

/* Generate some strings */
function POM_genAlphabet(isUpper = false) {
    return [...Array(26)].map((_, i) => String.fromCharCode(i + (isUpper ? 65 : 97))).join('');
}
function POM_genDigits() {
    return [...Array(10)].map((_, i) => String.fromCharCode(i + 48)).join('');
}
/* Distill string down to unique characters by using a set */
function POM_distillString(s) {
    s = s.split('');
    s = new Set(s);
    s = [...s].join('+');
    return(s);
}
/* Escape special regex characters in string, from MDN */
function POM_escapeRegExp(s) {
  return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'); // $& means the whole matched string
}
/* Validate "manual input" */
function POM_validateNumInput() {
    let mkLengthSlider = document.getElementById('mk-password-len-slider');
    if (mkLengthSlider) {
        if (this.value === '') {
            this.value = POM_minLength;
        }
        mkLengthSlider.value = this.value;
        if (mkLengthSlider.value < POM_minLength) {
            mkLengthSlider.value = POM_minLength;
        } else if (mkLengthSlider.value > POM_maxLength) {
            mkLengthSlider.value = POM_maxLength;
        }
        this.value = mkLengthSlider.value;
    }
    if (POM_changeIsClick) {
        document.getElementById('mk-password-btn').click();
    }
}
/* Validate "manual input" */
function POM_generatePassword() {
    let strengthP = document.getElementById('mk-password-strength');
    strengthP.style.background = 'inherit';
    strengthP.style.width = '0';
    strengthP.style.transition = 'none';
    let passwordField = document.getElementById('mk-password-field');
    passwordField.value = '';
    passwordField.classList.remove('mk-password-field-focus');
    let mkLengthField = document.getElementById('mk-length-field');
    if (POM_debug) {
        console.clear();
    }
    if (mkLengthField) {
        if (mkLengthField.value >= POM_minLength && mkLengthField.value <= POM_maxLength) {
            let passwordSource = '';
            if (document.getElementById('mk-uppercase-select').checked) {
                passwordSource += POM_strUppercase;
            }
            if (document.getElementById('mk-special-one-select').checked) {
                passwordSource += (POM_strSpecialOne + POM_strSpecialOne);
            }
            if (document.getElementById('mk-digits-select').checked) {
                passwordSource += POM_strDigits;
            }
            if (document.getElementById('mk-lowercase-select').checked) {
                passwordSource += POM_strLowercase;
            }
            if (document.getElementById('mk-digits-select').checked) {
                passwordSource += POM_strDigits;
            }
            if (document.getElementById('mk-special-two-select').checked) {
                passwordSource += (POM_strSpecialTwo + POM_strSpecialTwo);
            }
            if (passwordSource.length>0) {
                let passwordGen = '';
                let theChar = '';
                for (let i = 0; i < mkLengthField.value; i++) {
                    const charPos = Math.floor(Math.random() * passwordSource.length);
                    theChar = passwordSource[charPos];
                    passwordGen += theChar;
                }
                passwordField.value = passwordGen;

                let scoreP = 0;/* Password strength score, max is 8 */
                let searchStr;
                let re;

                /* See if password contains uppercase characters */
                searchStr = '[' + POM_escapeRegExp(POM_strUppercase) + ']';
                if (POM_debug) {
                    console.log('Using ' + searchStr + ' to scan password');
                }
                re = new RegExp(searchStr);
                if (re.test(passwordGen)) {
                    scoreP++;
                    if (POM_debug) {
                        console.log('Password contains UPPERCASE');
                    }
                } else {
                    if (POM_debug) {
                        console.log('Password does NOT contain UPPERCASE');
                    }
                }
                /* See if password contains lowercase characters */
                searchStr = '[' + POM_escapeRegExp(POM_strLowercase) + ']';
                if (POM_debug) {
                    console.log('Using ' + searchStr + ' to scan password');
                }
                re = new RegExp(searchStr);
                if (re.test(passwordGen)) {
                    scoreP++;
                    if (POM_debug) {
                        console.log('Password contains lowercase');
                    }
                } else {
                    if (POM_debug) {
                        console.log('Password does NOT contain lowercase');
                    }
                }
                /* See if password contains digits characters */
                searchStr = '[' + POM_escapeRegExp(POM_strDigits) + ']';
                if (POM_debug) {
                    console.log('Using ' + searchStr + ' to scan password');
                }
                re = new RegExp(searchStr);
                if (re.test(passwordGen)) {
                    scoreP++;
                    if (POM_debug) {
                        console.log('Password contains digits');
                    }
                } else {
                    if (POM_debug) {
                        console.log('Password does NOT contain digits');
                    }
                }
                /* See if password contains specialOne characters */
                searchStr = '[' + POM_escapeRegExp(POM_strSpecialOne) + ']';
                if (POM_debug) {
                    console.log('Using ' + searchStr + ' to scan password');
                }
                re = new RegExp(searchStr);
                if (re.test(passwordGen)) {
                    scoreP++;
                    if (POM_debug) {
                        console.log('Password contains specialOne');
                    }
                } else {
                    if (POM_debug) {
                        console.log('Password does NOT contain specialOne');
                    }
                }
                /* See if password contains specialtwo characters */
                searchStr = '[' + POM_escapeRegExp(POM_strSpecialTwo) + ']';
                if (POM_debug) {
                    console.log('Using ' + searchStr + ' to scan password');
                }
                re = new RegExp(searchStr);
                if (re.test(passwordGen)) {
                    scoreP++;
                    if (POM_debug) {
                        console.log('Password contains specialTwo');
                    }
                } else {
                    if (POM_debug) {
                        console.log('Password does NOT contain specialTwo');
                    }
                }
                /* Bump score if password is > 7 characters */
                if (passwordGen.length > 7) {
                    scoreP++;
                    if (POM_debug) {
                        console.log('Password >7 characters, buping score');
                    }
                }
                /* Bump score if password is > 15 characters */
                if (passwordGen.length > 15) {
                    scoreP++;
                    if (POM_debug) {
                        console.log('Password >15 characters, buping score');
                    }
                }
                /* Bump score if password is > 31 characters */
                if (passwordGen.length > 31) {
                    scoreP += 2;
                    if (POM_debug) {
                        console.log('Password >31 characters, buping score');
                    }
                }
                /* Bump score if password is > 63 characters */
                if (passwordGen.length > 63) {
                    scoreP += 4;
                    if (POM_debug) {
                        console.log('Password >63 characters, buping score');
                    }
                }
                /* Let's not overflooooow */
                if (scoreP > 8) {
                    scoreP = 8;
                }
                if (scoreP > 6) {
                    strengthP.style.background = '#27ce60';
                } else if (scoreP > 5) {
                    strengthP.style.background = '#27ae60';
                } else if (scoreP > 3) {
                    strengthP.style.background = '#FFC300';
                } else {
                    strengthP.style.background = '#C0392B';
                }
                strengthP.style.width = (scoreP * 12.5) + '%';
                strengthP.style.transition = 'width 0.3s ease-in';
            }
        }
    }
}
/* Copy content of generated password field, if any, to clipboard */
function POM_copyPassword() {
    let passwordField = document.getElementById('mk-password-field');
    if (passwordField && passwordField.value.length > 0) {
        if (navigator && navigator.clipboard && navigator.clipboard.writeText) {
            /* Use "modern" method */
            navigator.clipboard.writeText(passwordField.value).then(function() {
                passwordField.classList.add('mk-password-field-focus');
            }, function(e) {
                alert("Could not copy text: " + e);
            });
        } else {
            /* Use legacy method */
            var theText = document.createElement('textarea');
            theText.value = passwordField.value;
            theText.setAttribute('readonly', '');
            theText.style.position = 'absolute';
            theText.style.top = '-6969px';
            theText.style.left = '-6969px';
            document.body.appendChild(theText);
            theText.focus();
            theText.select();
            try {
                var goodCopy = document.execCommand('copy');
                if (goodCopy) {
                    passwordField.classList.add('mk-password-field-focus');
                } else {
                    alert('That did not work');
                }
            } catch(e) {
                alert('Could not copy text: ' + e.message);
            }
            document.body.removeChild(theText);
        }
    }
}
/* Initialize */
function POM_initialSetup() {
    /* Setup, language strings, etc. Ugly, but it works :-) */
    if (navigator.language) {
        POM_cfgLanguage = navigator.language;
        if (POM_debug) {
            console.log("Language: " + POM_cfgLanguage);
        }
        let dashPos = POM_cfgLanguage.indexOf('-');
        if (dashPos && dashPos>0) {
            POM_cfgLanguage = POM_cfgLanguage.substring(0, dashPos);
        }
        POM_cfgLanguage = POM_cfgLanguage.toLowerCase();
    } else {
        POM_cfgLanguage = '???';
    }
    /* Add your translation here */
    switch(POM_cfgLanguage) {
        case 'sv':
        case 'se':
            POM_cfgTitle = 'Slumpmässigt Lösenord';
            POM_cfgSlogan = 'Inga kakor, <span tyle="display:inline-block">ingen spårning.</span><br/>Bara lösenord.';
            POM_cfgGenPasswordHint = 'Generera lösenord';
            break;
        case 'de':
            /* Thank you Peter Hampf */
            POM_cfgTitle = 'Zufallsgenerator für Passwörter';
            POM_cfgSlogan = 'Keine Cookies, <span style="display:inline-block">keine Tracker.</span><br/>Nur Passwörter.';
            POM_cfgGenPasswordHint = 'Passwort generieren';
            break;
        case 'hu':
            /* Thank you Mihaly Balassy */
            POM_cfgTitle = 'Véletlenszerű jelszógenerátor';
            POM_cfgSlogan = 'Cookie-k nélkül, <span style="display:inline-block">nyomkövetők nélkül.</span><br/>Csak jelszavak.';
            POM_cfgGenPasswordHint = 'Jelszó generálása';
            break;
        case 'nl':
            /* Thank you Jeroen van de Leur */
            POM_cfgTitle = 'Willekeurig Wachtwoord Generator';
            POM_cfgSlogan = 'Geen cookies, <span style="display:inline-block">geen trackers.</span><br/>Uitsluitend wachtwoorden.';
            POM_cfgGenPasswordHint = 'Genereer wachtwoord';
            break;
        case 'lb':
            /* Thank you Alain Fontaine */
            POM_cfgTitle = 'Zoufallspasswuertgenerator';
            POM_cfgSlogan = 'Keng Cookien, <span style="display:inline-block">keng Trackeren.</span><br/>Nemme Passwierder.';
            POM_cfgGenPasswordHint = 'Passwuert genereieren';
            break;
        case 'fr':
            /* Thank you Alain Fontaine */
            POM_cfgTitle = 'Générateur de mot de passe au hazard';
            POM_cfgSlogan = 'Pas de cookies, <span style="display:inline-block">pas de traceurs.</span><br/>Juste des mots de passe.';
            POM_cfgGenPasswordHint = 'Générer mot de passe';
            break;
        case 'fi':
            /* Thank you Thomas Raehalme */
            POM_cfgTitle = 'Salasanageneraattori';
            POM_cfgSlogan = 'Ei evästeitä, <span style="display:inline-block">ei seurantaa.</span><br/>Vain salasanoja.';
            POM_cfgGenPasswordHint = 'Generoi salasana';
            break;
        case 'sl':
            /* Thank you Gregor Godler */
            POM_cfgTitle = 'Generator Naključnih Gesel';
            POM_cfgSlogan = 'Brez piškotkov, <span style="display:inline-block">nič sledilnikov.</span><br/>Samo geslo.';
            POM_cfgGenPasswordHint = 'Generiraj geslo';
            break;
        default:
            POM_cfgTitle = 'Random Password Generator';
            POM_cfgSlogan = 'No cookies, <span style="display:inline-block">no trackers.</span><br/>Just passwords.';
            POM_cfgGenPasswordHint = 'Generate password';
            break;
    }
    /* Paint "HTML" with language strings */
    document.getElementById('cfg-page-title').innerHTML = POM_cfgTitle;
    document.getElementById('cfg-page-slogan').innerHTML = POM_cfgSlogan;
    document.getElementById('cfg-page-lang').innerHTML = '[' + POM_cfgLanguage + ']';
    document.getElementById('mk-password-btn').title = POM_cfgGenPasswordHint;
    /* Set reasonable defaults */
    document.getElementById('mk-uppercase-select').checked = POM_strUppercase_Default;
    document.getElementById('mk-lowercase-select').checked = POM_strLowercase_Default;
    document.getElementById('mk-digits-select').checked = POM_strDigits_Default;
    document.getElementById('mk-special-one-select').checked = POM_strSpecialOne_Default;
    document.getElementById('mk-special-two-select').checked = POM_strSpecialTwo_Default;

    document.getElementById('mk-password-field').value = '';
    let mkPasswordSlider = document.getElementById('mk-password-len-slider');
    if (mkPasswordSlider) {
        mkPasswordSlider.value = POM_minLength;
        mkPasswordSlider.setAttribute("min", POM_minLength);
        mkPasswordSlider.setAttribute("max", POM_maxLength);
        mkPasswordSlider.step = POM_sliderStep;
    }
    document.getElementById('mk-length-field').value = POM_minLength;
    /* Paint "HTML" with password source */
    document.getElementById("gen-uppercase").innerText = POM_strUppercase;
    document.getElementById("gen-lowercase").innerText = POM_strLowercase;
    document.getElementById("gen-digits").innerText = POM_strDigits;
    document.getElementById("gen-special-one").innerText = POM_strSpecialOne;
    document.getElementById("gen-special-two").innerText = POM_strSpecialTwo;
    /* Add timestamp, for no good reason :) */
    document.getElementById("gen-timestamp").innerText = Date.now();
    /* Add some event handlers */
    document.getElementById('mk-password-len-slider').addEventListener('input', function() {
        let mkLengthField = document.getElementById('mk-length-field');
        if (mkLengthField) {
            mkLengthField.value = this.value;
            if (mkLengthField.value < POM_minLength) {
                mkLengthField.value = POM_minLength;
            } else if (mkLengthField.value > POM_maxLength) {
                mkLengthField.value = POM_maxLength;
            }
            this.value = mkLengthField.value;
        }
    });
    document.getElementById('mk-length-field').addEventListener('click', function() {
        this.select();
    });
    document.getElementById('mk-length-field').addEventListener('blur', POM_validateNumInput);
    document.getElementById('mk-length-field').addEventListener('change', POM_validateNumInput);
    document.getElementById('mk-password-btn').addEventListener('click', POM_generatePassword);
    document.getElementById('mk-password-field').addEventListener('click', POM_copyPassword);
    /* Possibly generate password on load */
    if (POM_mkPasswordOnLoad) {
        document.getElementById('mk-password-btn').click();
    }
    /* Possibly generate password on configuration changes */
    if (POM_changeIsClick) {
        document.getElementById('mk-uppercase-select').addEventListener('click', POM_generatePassword);
        document.getElementById('mk-special-one-select').addEventListener('click', POM_generatePassword);
        document.getElementById('mk-digits-select').addEventListener('click', POM_generatePassword);
        document.getElementById('mk-lowercase-select').addEventListener('click', POM_generatePassword);
        document.getElementById('mk-digits-select').addEventListener('click', POM_generatePassword);
        document.getElementById('mk-special-two-select').addEventListener('click', POM_generatePassword);
        document.getElementById('mk-password-len-slider').addEventListener('change', POM_generatePassword);
    }

}


/* Wait for things to be loaded */
if (document.readyState === "complete" ||
        (document.readyState !== "loading" && !document.documentElement.doScroll)) {
    POM_initialSetup();
} else {
    document.addEventListener("DOMContentLoaded", POM_initialSetup);
}

/* end of file "password-om.js" */
